app/src/main/java/ai/elimu/appstore/BaseApplication.java

```diff
@@ -92,13 +102,55 @@ public Retrofit getRetrofit() {
         HttpLoggingInterceptor logging = new HttpLoggingInterceptor();
         logging.setLevel(BuildConfig.DEBUG ? HttpLoggingInterceptor.Level.BODY :
                 HttpLoggingInterceptor.Level.NONE);
-        final OkHttpClient okHttpClient = new OkHttpClient.Builder()
+
+        final ProgressResponseBody.ProgressListener progressListener = new ProgressResponseBody.ProgressListener() {
+            @Override
+            public void update(long bytesRead, long contentLength, boolean done) {
+//                Timber.i("bytesRead: " + bytesRead);
+//                Timber.i("contentLength: " + contentLength);
+//                Timber.i("done: " + done);
+//                Timber.i("%d%% done\n", (100 * bytesRead) / contentLength);
+
+                if (progressUpdateCallback != null) {
+                    long progress = (bytesRead * 100) / contentLength;
+//                    Timber.d("progress: " + progress);
+
+                    // E.g. "6.00 MB/12.00 MB   50%"
+                    String progressText = String.format(context.getString(R.string
+                                    .app_list_download_progress_number), bytesRead / 1024f / 1024f,
+                            contentLength / 1024f / 1024f, progress);
+
+//                    Timber.i("progressText: " + progressText);
+                    progressUpdateCallback.onProgressUpdated(progressText, (int) progress);
+                }
+            }
+        };
+
+
+        final OkHttpClient.Builder okHttpClientBuilder = new OkHttpClient.Builder()
                 .addInterceptor(logging)
                 .readTimeout(20, TimeUnit.SECONDS)
-                .retryOnConnectionFailure(true)
-                .build();
+                .retryOnConnectionFailure(true);
```

Comment:
How many times will the retry happen? Once?

Reply:
Basically since we do not have many concurrent network requests, the maximum retry is 1.
In case we have many concurrent requests, then it depends on the thread pool from which requests are executed. Default number of retry in this case is 5.
Kindly refer to below comment

https://stackoverflow.com/a/47544627/3682198